src/extensions/codemirror/CodeMirror.tsx

```diff
@@ -119,23 +143,39 @@ export const CodeEditor = ({
                     cm.execCommand(cm.getOption("indentWithTabs") ? "insertTab" : "insertSoftTab");
                 },
             },
+            foldGutter: supportCodeFolding,
+            gutters: supportCodeFolding ? ["CodeMirror-linenumbers", "CodeMirror-foldgutter"] : [],
         });
 
-        editorInstance.on("change", (api) => {
-            onChange(api.getValue());
-        });
+        setEditorInstance && setEditorInstance(editorInstance);
+
+        if (onScroll) {
+            editorInstance.on("scroll", (instance) => {
+                onScroll(instance);
+            });
+        }
+
+        if (onChange) {
+            editorInstance.on("change", (api) => {
+                onChange(api.getValue());
+            });
+        }
 
         if (height) {
             editorInstance.setSize(null, height);
         }
 
+        editorInstance.setValue(defaultValue);
+
         return function cleanup() {
             editorInstance.toTextArea();
         };
     }, [onChange, mode, preventLineNumbers]);
 
+    const jsonModifierClassName = mode === "json" ? `${eccgui}-json-modifier` : "";
```

Comment:
Class name not valid against BEM guidelines, should probably be `${eccgui}-codeeditor--mode-json`. And this class should be set for all modes (then probably `${eccgui}-codeeditor--mode-${mode}`), even if we only use it for `json` later in the stylesheets.